python/plugins/processing/algs/qgis/RasterSampling.py

```diff
@@ -0,0 +1,163 @@
+
+    def displayName(self):
+        return self.tr('Sample Raster Values')
+
+    def group(self):
+        return self.tr('Raster analysis')
+
+    def groupId(self):
+        return 'rasteranalysis'
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(
+            QgsProcessingParameterFeatureSource(
+                self.INPUT,
+                self.tr('Input Point Layer'),
+                [QgsProcessing.TypeVectorPoint]
+            )
+        )
+
+        self.addParameter(
+            QgsProcessingParameterMultipleLayers(
+                self.RASTERCOPY,
+                self.tr('Raster Layer to sample'),
+                QgsProcessing.TypeRaster
+            )
+        )
+
+        self.addParameter(
+            QgsProcessingParameterFeatureSink(
+                self.OUTPUT,
+                self.tr('Sampled Points')
+            )
+        )
+
+    def processAlgorithm(self, parameters, context, feedback):
+
+        source = self.parameterAsSource(
+            parameters,
+            self.INPUT,
+            context
+        )
+
+        sampled_rasters = self.parameterAsLayerList(
+            parameters,
+            self.RASTERCOPY,
+            context
+        )
+
+        if source is None:
+            raise QgsProcessingException(self.invalidSourceError(parameters, self.INPUT))
+
+        source_fields = source.fields()
+        raster_fields = QgsFields()
+
+        # append field to vector as rasterName_bandCount
+        for i in sampled_rasters:
+            for b in range(i.bandCount()):
+                raster_fields.append(QgsField(
+                    os.path.splitext(os.path.basename(i.source()))[0] + str('_{}'.format(b + 1)), QVariant.Double)
+                )
+
+        # combine all the vector fields
+        out_fields = QgsProcessingUtils.combineFields(source_fields, raster_fields)
+
+        (sink, dest_id) = self.parameterAsSink(
+            parameters,
+            self.OUTPUT,
+            context,
+            out_fields,
+            source.wkbType(),
+            source.sourceCrs()
+        )
+
+        if sink is None:
+            raise QgsProcessingException(self.invalidSinkError(parameters, self.OUTPUT))
+
+        total = 100.0 / source.featureCount() if source.featureCount() else 0
+        features = source.getFeatures()
+
+        for n, i in enumerate(source.getFeatures()):
+
+            for rr in sampled_rasters:
+
+                attrs = i.attributes()
+
+                if rr.bandCount() > 1:
+
+                    for b in range(rr.bandCount()):
+                        attrs.append(
+                            rr.dataProvider().identify(i.geometry().asPoint(),
+                                                       QgsRaster.IdentifyFormatValue).results()[b + 1]
+                        )
+
+                attrs.append(
+                    rr.dataProvider().identify(i.geometry().asPoint(),
```

Comment:
@wonder-sk How well do you know the `identify` method? Do you know if there's a more efficient way to do this sampling?

Reply:
Actually it's not very efficient - to much other stuff happening in that method. I'll add an optimised "sample" method which ONLY does the raster sampling and none of the other identification steps.

Reply:
@nyalldawson IMHO this approach is a bit slow. Years ago I developed script which processes 2500 points on singleband raster much faster, in 0,1725 sec

Reply:
Ok, see https://github.com/qgis/QGIS/pull/7403 for an alternative approach